client/manager/manager.go

```diff
@@ -0,0 +1,154 @@
+	SecretGenerator   func() ([]byte, error)
+	ClientIDGenerator func(string) (string, error)
+}
+
+func NewClientManager(clientRepo client.ClientRepo, txnFactory repo.TransactionFactory, options ManagerOptions) *ClientManager {
+	if options.SecretGenerator == nil {
+		options.SecretGenerator = secret.DefaultSecretGenerator
+	}
+	if options.ClientIDGenerator == nil {
+		options.ClientIDGenerator = oidc.GenClientID
+	}
+	return &ClientManager{
+		clientRepo:        clientRepo,
+		begin:             txnFactory,
+		secretGenerator:   options.SecretGenerator,
+		clientIDGenerator: options.ClientIDGenerator,
+	}
+}
+
+func (m *ClientManager) New(meta oidc.ClientMetadata) (*oidc.ClientCredentials, error) {
+	tx, err := m.begin()
+	if err != nil {
+		return nil, err
+	}
+	defer rollback(tx)
+
+	// Generate Client ID
+	clientID, err := m.clientIDGenerator(meta.RedirectURIs[0].Host)
+
+	// Generate Secret
+	secret, err := m.secretGenerator()
+	if err != nil {
+		return nil, err
+	}
+	clientSecret := base64.URLEncoding.EncodeToString(secret)
+	cli := client.Client{
+		Credentials: oidc.ClientCredentials{
+			ID:     clientID,
+			Secret: clientSecret,
+		},
+		Metadata: meta,
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	// Save Client
+	cc, err := m.clientRepo.New(tx, cli)
+	if err != nil {
+		return nil, err
+	}
+
+	err = tx.Commit()
+	if err != nil {
+		return nil, err
+	}
+	return cc, nil
+}
+
+func (m *ClientManager) Get(id string) (client.Client, error) {
+	return m.clientRepo.Get(nil, id)
+}
+
+func (m *ClientManager) All() ([]client.Client, error) {
+	return m.clientRepo.All(nil)
+}
+
+func (m *ClientManager) Metadata(clientID string) (*oidc.ClientMetadata, error) {
+	c, err := m.clientRepo.Get(nil, clientID)
+	if err != nil {
+		return nil, err
+	}
+
+	return &c.Metadata, nil
+}
+
+func (m *ClientManager) IsDexAdmin(clientID string) (bool, error) {
+	c, err := m.clientRepo.Get(nil, clientID)
+	if err != nil {
+		return false, err
+	}
+
+	return c.Admin, nil
+}
+
+func (m *ClientManager) SetDexAdmin(clientID string, isAdmin bool) error {
+	tx, err := m.begin()
+	if err != nil {
+		return err
+	}
+	defer rollback(tx)
+
+	c, err := m.clientRepo.Get(tx, clientID)
+	if err != nil {
+		return err
+	}
+
+	c.Admin = isAdmin
+	err = m.clientRepo.Update(tx, c)
+	if err != nil {
+		return err
+	}
+
+	err = tx.Commit()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m *ClientManager) Authenticate(creds oidc.ClientCredentials) (bool, error) {
+	clientSecret, err := m.clientRepo.GetSecret(nil, creds.ID)
+	if err != nil || clientSecret == nil {
+		return false, nil
+	}
+
+	dec, err := base64.URLEncoding.DecodeString(creds.Secret)
+	if err != nil {
+		log.Errorf("error Decoding client creds: %v", err)
+		return false, nil
+	}
+
+	ok := secret.CompareHashAndPassword(clientSecret, dec) == nil
+	return ok, nil
+}
+
+func rollback(tx repo.Transaction) {
+	err := tx.Rollback()
+	if err != nil {
+		log.Errorf("unable to rollback: %v", err)
```

Comment:
Won't this trigger on every successful transaction because you defer the rollback call?